enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,756 @@
+
+### Implementation Details/Notes/Constraints
+
+#### Synchronisation of resources
+
+To ensure that resources are synchronised, we will implement a new controller `machine-api-synchronisation-controller`.
+This controller will be responsible for translation of Machine API resources to Cluster API resources, and vice-versa.
+
+This controller will be responsible for creating Cluster API versions of Machine API resources, when they do not exist.
+It will observe the authoritative annotation described [above][authoritative-annotation] to synchronise between the two resources,
+overwriting anything in the non-authoritative resource that does not match with the authoritative resource.
+
+When errors occur during synchronisation, the controller will write conditions to the Machine API resource to report the failure.
+
+On successful synchronisation, the controller will add an annotation to the non-authoritative resource to indicate the observed generation of the last successful synchronisation.
+The annotation `machine.openshift.io/synchronised-generation` will be set with a value of the generation of the authoritative resource, onto the non-authoritative resource.
+This will allow users to keep track of, and ensure that the latest changes to the authoritative resource have been synchronised.
+
+When the `machine.openshift.io/synchronised-generation` is present on the authoritative resource, it will be cleared during synchronisation.
+
+[authoritative-annotation](#authority-of-resources-and-controller-reconciliations)
+
+##### MachineSet Synchronisation
+
+MachineSets in Machine API map to a MachineSet and InfrastructureTemplate in Cluster API.
+The synchronisation controller will create the Cluster API MachineSet with the same name as the Machine API MachineSet to ensure a 1:1 mapping between these resources.
+The InfrastructureTemplate will be named based on the MachineSet name and a hash of the content within it [^6].
+InfrastructureTemplates are immutable, but MachineSet’s in Machine API are not, should the content of the MachineSet change,
+a new InfrastructureTemplate will be created and the old template will be removed when no longer required.
+
+The `providerSpec` forms the basis of the conversion to the InfrastructureTemplate in this case and is where errors may occur.
+Some values from the `providerSpec` in Machine API are moved to the InfrastructureCluster resource (a cluster wide infrastructure reference) and as such,
+if these values do not match that which is set on the InfrastructureCluster, we cannot convert losslessly.
+
+To understand the likelihood of this, we can use existing CCX data [^7] to determine if there will be clashes in this data before we implement the controller. 
+
+In cases where conversion fails, the controller will add a condition to the Machine API MachineSet defining the errors present.
```

Comment:
Seems like there are at least three failure modes here:
- The syncer's internal logic is unable to come up with what it considers a sane target resource. This one is easy: update the condition as noted.
- The syncer thinks it came up with a reasonable target resource, but gets an HTTP error when POSTing it to the server. This comes in at least two flavors:
  - Retryable, e.g. 409 conflict. Requeue and try again next time.
  - Fatal, e.g. rejected by webhook. For this one we probably also update the condition.
- This is the weird one: k8s accepts the POST, but the object's controller -- *if it ran* -- would have noticed some internal problem and marked its (non-this-enhancement-related) status conditions to indicate failure. We either:
  - Code the non-authoritative controller to essentially "dry run" the update and set its conditions accordingly. 
    - First, this would probably be Hard™, as not all of that logic is likely to be linear or in one place.
    - Second, how does the syncer know when that's done?
  - **OR** the non-authoritative controller truly no-ops. This is the option that seems correct based on this design. However, then the problem lies dormant and undetectable until a subsequent auth switch happens! Effectively the migration broke a Machine.
    ...but I guess this is okay: it means the migration didn't work, and the admin can/should just flip auth back to the original. And, I guess, iterate?
    It doesn't sound ideal, but I think I've convinced myself that this is the right option.

Reply:
Non-authoritative should truly no-op, and yes, there may well be broken machines. This is why we are introducing this initially with a get out of jail card, the ability to switch back